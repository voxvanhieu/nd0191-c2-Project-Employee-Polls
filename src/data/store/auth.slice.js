import { createSlice, createAsyncThunk, createReducer } from "@reduxjs/toolkit";

import * as API from '../api';
import { alertActions } from "./alert.slice";
import { history } from '../../helpers'

// Create slice
const name = 'auth';
const initialState = createInitialState();
const selectors = createSelectors();
const reducers = createReducers();
const extraActions = createExtraActions();
const slice = createSlice({ name, initialState, reducers, selectors });

// Exports
export const authActions = { ...slice.actions, ...extraActions };
export const authSelectors = { ...slice.selectors }
export const authReducer = slice.reducer;

// Implementation

function createInitialState() {
    return {
        value: JSON.parse(localStorage.getItem('auth'))
    }
}

function createSelectors() {
    return {
        selectValue: (sliceState) => sliceState.value,
    }
}

function createReducers() {

    return {
        setAuth
    }

    function setAuth(state, action) {
        state.value = action.payload;
    }
}

function createExtraActions() {

    return {
        login: _login(),
        logout: _logout()
    }

    function _login() {
        return createAsyncThunk(
            `${name}/login`,
            async ({ username, password }, { dispatch }) => {
                // dispatch(alertActions.clear());
                try {
                    let user = await API.authenticateUser({ username, password });

                    dispatch(authActions.setAuth(user));

                    localStorage.setItem('auth', JSON.stringify(user));

                    if (user && user.name) {
                        dispatch(alertActions.success("Welcome"));
                        setTimeout(() => {
                            const { from } = history.location.state || { from: { pathname: '/' } };
                            history.navigate(from);
                        }, 500);
                    } else {
                        dispatch(alertActions.error("Wrong username or password"));
                    }
                } catch (error) {
                    console.log(error);
                }
            }
        )
    }

    function _logout() {
        return createAsyncThunk(
            `${name}/logout`,
            function (arg, { dispatch }) {
                dispatch(authActions.setAuth(null));
                localStorage.removeItem('auth');
                history.navigate('/login');
            }
        )
    }
}