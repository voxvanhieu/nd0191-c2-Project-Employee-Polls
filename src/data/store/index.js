import { configureStore } from '@reduxjs/toolkit';

import { alertReducer } from './alert.slice';
import { authReducer } from './auth.slice'
import { userReducer as usersReducer } from './user.slice'
import { pollsReducer } from './poll.slice';

export * from './alert.slice'
export * from './auth.slice'
export * from './user.slice'
export * from './poll.slice'

export const store = configureStore({
    reducer: {
        alert: alertReducer,
        auth: authReducer,
        users: usersReducer,
        polls: pollsReducer,
    }
})