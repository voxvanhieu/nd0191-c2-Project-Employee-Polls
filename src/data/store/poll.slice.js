import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

import * as API from '../api';
import { useDispatch } from "react-redux";

// Create slice
const name = 'polls';
const initialState = createInitialState();
const selectors = createSelectors();
const advSelectors = createAdvancedSelectors();
const reducers = createReducers();
const extraActions = createExtraActions();
const extraReducers = createExtraReducers();

const slice = createSlice({ name, initialState, reducers, extraReducers, selectors });

// Exports
export const pollActions = { ...slice.actions, ...extraActions };
export const pollSelectors = { ...slice.selectors, ...advSelectors }
export const pollsReducer = slice.reducer;

// Implementation

function createInitialState() {
    return {
        allIds: [],
        byId: {},
        status: {}
    }
}

function createSelectors() {
    return {
        selectValue: (sliceState) => sliceState.byId,
    }
}

//  Selector with parameters
function createAdvancedSelectors() {
    return {
        selectQuestionById: (pollId) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.includes(pollId)
                ? sliceState.byId[pollId]
                : null;
        },
        selectQuestionIds: (pollIds) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.filter(key => pollIds.includes(key));
        },
        selectQuestionIdsExcepts: (pollIds) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.filter(key => !pollIds.includes(key));
        }
    }
}


function createReducers() {
    return {
        getById: {
            reducer: (state, action) => state.byId[action.payload.id],
            prepare: (id) => ({ payload: id })
        },
    };
}

function createExtraActions() {

    return {
        getQuestions: createAsyncThunk(
            `${name}/getQuestions`,
            async () => await API._getQuestions()
        ),
    }
}

function createExtraReducers() {
    return (builder) => {
        buildReducerGetQuestions();

        function buildReducerGetQuestions() {
            var { pending, fulfilled, rejected } = extraActions.getQuestions;
            builder
                .addCase(pending, (state) => {
                    state.status = { loading: true };
                })
                .addCase(fulfilled, (state, action) => {
                    state.byId = action.payload;
                    state.allIds = Object.keys(state.byId);
                    state.status = { loading: false };
                })
                .addCase(rejected, (state, action) => {
                    state.status = { error: action.error };
                })
        }
    }
}
