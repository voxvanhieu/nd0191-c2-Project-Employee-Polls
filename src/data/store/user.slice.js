import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

import * as API from '../api';

// Create slice
const name = 'users';
const initialState = createInitialState();
const selectors = createSelectors();
const advSelectors = createAdvancedSelectors();
const extraActions = createExtraActions();
const extraReducers = createExtraReducers();
const slice = createSlice({ name, initialState, extraReducers, selectors });

// Exports
export const userActions = { ...slice.actions, ...extraActions };
export const userSelectors = { ...slice.selectors, ...advSelectors };
export const userReducer = slice.reducer;

// Implementation

function createInitialState() {
    return {
        allIds: [],
        byId: {},
        status: {}
    }
}

function createSelectors() {
    return {
        selectValue: (sliceState) => sliceState.byId,
    }
}

//  Selector with parameters
function createAdvancedSelectors() {
    return {
        selectUserInfoById: (userId) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.includes(userId) ? sliceState.byId[userId] : null;
        },
        selectCompletedPolls: (userId) => (state) => {
            const sliceState = state[slice.name];
            return sliceState.allIds.includes(userId) ? sliceState.byId[userId].answers : null;
        }
    }
}

function createExtraActions() {

    return {
        getUsers: createAsyncThunk(
            `${name}/getUsers`,
            async () => await API._getUsers()
        )
    }
}

function createExtraReducers() {
    return (builder) => {
        buildReducerGetUsersAsync();

        function buildReducerGetUsersAsync() {
            var { pending, fulfilled, rejected } = extraActions.getUsers;
            builder
                .addCase(pending, (state) => {
                    state.status = { loading: true };
                })
                .addCase(fulfilled, (state, action) => {
                    state.byId = action.payload;
                    state.allIds = Object.keys(action.payload);
                    state.status = { loading: false };
                })
                .addCase(rejected, (state, action) => {
                    state.status = {
                        loading: false,
                        error: action.error
                    };
                })
        }
    }
}