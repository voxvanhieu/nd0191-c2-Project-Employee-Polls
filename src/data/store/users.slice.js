import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

import * as API from '../api';

// Create slice
const name = 'users';
const initialState = createInitialState();
const extraActions = createExtraActions();
const extraReducers = createExtraReducers();
const slice = createSlice({ name, initialState, extraReducers });

// Exports
export const userAction = { ...slice.actions, ...extraActions };
export const usersReducer = slice.reducer;

// Implementation

function createInitialState() {
    return {
        list: null,
        item: null
    }
}

function createExtraActions() {

    return {
        register: _register(),
        getAll: _getAll(),
        getByUsername: _getByUsername(),
    }

    function _register() {
        // TBD
        return createAsyncThunk(
            `${name}/register`,
            async (user) => await new Promise((resolve) => {
                setTimeout(() => {
                    resolve(user);
                }, 1000);
            })
        );
    }

    function _getAll() {
        return createAsyncThunk(
            `${name}/getAll`,
            async () => await API._getUsers()
        );
    }

    function _getByUsername() {
        return createAsyncThunk(
            `${name}/getById`,
            async (id) => await API._getUserByUsername(id)
        );
    }
}

function createExtraReducers() {
    return (builder) => {
        buildReducerGetAll();
        buildReducerGetByUsername();

        function buildReducerGetAll() {
            var { pending, fulfilled, rejected } = extraActions.getAll;
            builder
                .addCase(pending, (state) => {
                    state.list = { loading: true };
                })
                .addCase(fulfilled, (state, action) => {
                    state.list = { value: action.payload };
                })
                .addCase(rejected, (state, action) => {
                    state.list = { error: action.error };
                })
        }

        function buildReducerGetByUsername() {
            var { pending, fulfilled, rejected } = extraActions.getByUsername;
            builder
                .addCase(pending, (state) => {
                    state.item = { loading: true };
                })
                .addCase(fulfilled, (state, action) => {
                    state.item = { value: action.payload };
                })
                .addCase(rejected, (state, action) => {
                    state.item = { error: action.error };
                })
        }
    }
}