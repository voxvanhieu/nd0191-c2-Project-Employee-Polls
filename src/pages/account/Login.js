import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { useForm } from "react-hook-form";
import { yupResolver } from '@hookform/resolvers/yup';
import * as Yup from 'yup';

import { authActions } from '../../data/store';

export { Login };

function Login() {
    const dispatch = useDispatch();

    // form validation rules 
    const validationSchema = Yup.object().shape({
        username: Yup.string().required('Username is required'),
        password: Yup.string().required('Password is required')
    });
    const formOptions = { resolver: yupResolver(validationSchema) };

    // get functions to build form with useForm() hook
    const { register, handleSubmit, formState } = useForm(formOptions);
    const { errors, isSubmitting } = formState;

    function onSubmit({ username, password }) {
        return dispatch(authActions.login({ username, password }));
    }

    return (
        <div className="container mx-auto">
            <div className="flex flex-col items-center px-6 py-8 mx-auto md:h-screen lg:py-0">
                <a href="#" className="flex items-center mb-6 text-2xl font-semibold text-gray-900 dark:text-white">
                    <img className="w-8 h-8 mr-2" src="uda-ico.svg" alt="logo" />
                    Udacity Employee Poll
                </a>
                <div className="w-full bg-white rounded-lg shadow dark:border md:mt-0 sm:max-w-md xl:p-0 dark:bg-gray-800 dark:border-gray-700">
                    <div className="p-6 space-y-4 md:space-y-6 sm:p-8">
                        <h1 className="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl dark:text-white">
                            Sign in to your account
                        </h1>
                        <form className="space-y-4 md:space-y-6" onSubmit={handleSubmit(onSubmit)}>
                            <div>
                                <label htmlFor="username" data-testid="lblUsername" className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">User Name</label>
                                <input type="text"
                                    data-testid="inpUsername"
                                    name="username" {...register('username')}
                                    className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                                    placeholder="employee_account" />
                            </div>
                            <div>
                                <label htmlFor="password" data-testid="lblPassword" className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">Password</label>
                                <input type="password"
                                    data-testid="inpPassword"
                                    name="password" {...register('password')}
                                    className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                                    placeholder="••••••••" />
                            </div>
                            <div className="flex items-center justify-between">
                                <div className="flex items-start">
                                    <div className="flex items-center h-5">
                                        <input id="remember" aria-describedby="remember" type="checkbox" className="w-4 h-4 border border-gray-300 rounded bg-gray-50 focus:ring-3 focus:ring-primary-300 dark:bg-gray-700 dark:border-gray-600 dark:focus:ring-primary-600 dark:ring-offset-gray-800" required="" />
                                    </div>
                                    <div className="ml-3 text-sm">
                                        <label htmlFor="remember" className="text-gray-500 dark:text-gray-300">Remember me</label>
                                    </div>
                                </div>
                                <Link to={`/forgot`} className="text-sm font-medium text-primary-600 hover:underline dark:text-primary-500">Forgot password?</Link>
                            </div>
                            <button
                                disabled={isSubmitting}
                                type="submit"
                                data-testid="btnSubmit"
                                className="w-full text-white bg-primary-600 hover:bg-primary-700 focus:ring-4 focus:outline-none focus:ring-primary-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center dark:bg-primary-600 dark:hover:bg-primary-700 dark:focus:ring-primary-800">
                                Sign in
                            </button>
                            <p className="text-sm font-light text-gray-500 dark:text-gray-400">
                                Don't have an account yet?&nbsp;
                                <Link to={`/register`} className='font-medium text-primary-600 hover:underline dark:text-primary-500'>Sign up</Link>
                            </p>
                        </form>
                    </div>
                </div>
            </div>
        </div>
    )
}
