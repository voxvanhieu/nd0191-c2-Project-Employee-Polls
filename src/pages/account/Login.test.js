
import React from 'react';
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { act, fireEvent, render, screen, waitFor } from "@testing-library/react";
import { store } from '../../data/store';
import { Login } from './Login';


describe("Login page test", () => {
    it("should render the component", () => {
        const view = render(
            <React.StrictMode>
                <Provider store={store}>
                    <BrowserRouter>
                        <Login />
                    </BrowserRouter>
                </Provider>
            </React.StrictMode>
        );
        expect(view).toBeDefined();
        expect(view).toMatchSnapshot();
    })


    it("should rendder all elements and fire events", () => {
        const view = render(
            <React.StrictMode>
                <Provider store={store}>
                    <BrowserRouter>
                        <Login />
                    </BrowserRouter>
                </Provider>
            </React.StrictMode>
        );

        const lblUsername = screen.getByTestId("lblUsername");
        const inpUsername = screen.getByTestId("inpUsername");
        const lblPassword = screen.getByTestId("lblPassword");
        const inpPassword = screen.getByTestId("inpPassword");
        const btnSubmit = screen.getByTestId("btnSubmit");

        expect(lblUsername.textContent).toBe("User Name");
        expect(lblPassword.textContent).toBe("Password");
        expect(btnSubmit.textContent).toBe("Sign in");

        fireEvent.change(inpUsername, { target: { value: "tylermcginnis" } });
        fireEvent.change(inpPassword, { target: { value: "abc321" } });

        expect(inpUsername.value).toBe("tylermcginnis");
        expect(inpPassword.value).toBe("abc321");
    })
})