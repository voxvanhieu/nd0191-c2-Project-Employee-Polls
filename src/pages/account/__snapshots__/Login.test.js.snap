// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page test should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container mx-auto"
      >
        <div
          class="flex flex-col items-center px-6 py-8 mx-auto md:h-screen lg:py-0"
        >
          <a
            class="flex items-center mb-6 text-2xl font-semibold text-gray-900 dark:text-white"
            href="#"
          >
            <img
              alt="logo"
              class="w-8 h-8 mr-2"
              src="uda-ico.svg"
            />
            Udacity Employee Poll
          </a>
          <div
            class="w-full bg-white rounded-lg shadow dark:border md:mt-0 sm:max-w-md xl:p-0 dark:bg-gray-800 dark:border-gray-700"
          >
            <div
              class="p-6 space-y-4 md:space-y-6 sm:p-8"
            >
              <h1
                class="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl dark:text-white"
              >
                Sign in to your account
              </h1>
              <form
                class="space-y-4 md:space-y-6"
              >
                <div>
                  <label
                    class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                    data-testid="lblUsername"
                    for="username"
                  >
                    User Name
                  </label>
                  <input
                    class="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                    data-testid="inpUsername"
                    name="username"
                    placeholder="employee_account"
                    type="text"
                  />
                </div>
                <div>
                  <label
                    class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                    data-testid="lblPassword"
                    for="password"
                  >
                    Password
                  </label>
                  <input
                    class="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                    data-testid="inpPassword"
                    name="password"
                    placeholder="••••••••"
                    type="password"
                  />
                </div>
                <div
                  class="flex items-center justify-between"
                >
                  <div
                    class="flex items-start"
                  >
                    <div
                      class="flex items-center h-5"
                    >
                      <input
                        aria-describedby="remember"
                        class="w-4 h-4 border border-gray-300 rounded bg-gray-50 focus:ring-3 focus:ring-primary-300 dark:bg-gray-700 dark:border-gray-600 dark:focus:ring-primary-600 dark:ring-offset-gray-800"
                        id="remember"
                        type="checkbox"
                      />
                    </div>
                    <div
                      class="ml-3 text-sm"
                    >
                      <label
                        class="text-gray-500 dark:text-gray-300"
                        for="remember"
                      >
                        Remember me
                      </label>
                    </div>
                  </div>
                  <a
                    class="text-sm font-medium text-primary-600 hover:underline dark:text-primary-500"
                    href="/forgot"
                  >
                    Forgot password?
                  </a>
                </div>
                <button
                  class="w-full text-white bg-primary-600 hover:bg-primary-700 focus:ring-4 focus:outline-none focus:ring-primary-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center dark:bg-primary-600 dark:hover:bg-primary-700 dark:focus:ring-primary-800"
                  data-testid="btnSubmit"
                  type="submit"
                >
                  Sign in
                </button>
                <p
                  class="text-sm font-light text-gray-500 dark:text-gray-400"
                >
                  Don't have an account yet? 
                  <a
                    class="font-medium text-primary-600 hover:underline dark:text-primary-500"
                    href="/register"
                  >
                    Sign up
                  </a>
                </p>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container mx-auto"
    >
      <div
        class="flex flex-col items-center px-6 py-8 mx-auto md:h-screen lg:py-0"
      >
        <a
          class="flex items-center mb-6 text-2xl font-semibold text-gray-900 dark:text-white"
          href="#"
        >
          <img
            alt="logo"
            class="w-8 h-8 mr-2"
            src="uda-ico.svg"
          />
          Udacity Employee Poll
        </a>
        <div
          class="w-full bg-white rounded-lg shadow dark:border md:mt-0 sm:max-w-md xl:p-0 dark:bg-gray-800 dark:border-gray-700"
        >
          <div
            class="p-6 space-y-4 md:space-y-6 sm:p-8"
          >
            <h1
              class="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl dark:text-white"
            >
              Sign in to your account
            </h1>
            <form
              class="space-y-4 md:space-y-6"
            >
              <div>
                <label
                  class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                  data-testid="lblUsername"
                  for="username"
                >
                  User Name
                </label>
                <input
                  class="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                  data-testid="inpUsername"
                  name="username"
                  placeholder="employee_account"
                  type="text"
                />
              </div>
              <div>
                <label
                  class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                  data-testid="lblPassword"
                  for="password"
                >
                  Password
                </label>
                <input
                  class="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                  data-testid="inpPassword"
                  name="password"
                  placeholder="••••••••"
                  type="password"
                />
              </div>
              <div
                class="flex items-center justify-between"
              >
                <div
                  class="flex items-start"
                >
                  <div
                    class="flex items-center h-5"
                  >
                    <input
                      aria-describedby="remember"
                      class="w-4 h-4 border border-gray-300 rounded bg-gray-50 focus:ring-3 focus:ring-primary-300 dark:bg-gray-700 dark:border-gray-600 dark:focus:ring-primary-600 dark:ring-offset-gray-800"
                      id="remember"
                      type="checkbox"
                    />
                  </div>
                  <div
                    class="ml-3 text-sm"
                  >
                    <label
                      class="text-gray-500 dark:text-gray-300"
                      for="remember"
                    >
                      Remember me
                    </label>
                  </div>
                </div>
                <a
                  class="text-sm font-medium text-primary-600 hover:underline dark:text-primary-500"
                  href="/forgot"
                >
                  Forgot password?
                </a>
              </div>
              <button
                class="w-full text-white bg-primary-600 hover:bg-primary-700 focus:ring-4 focus:outline-none focus:ring-primary-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center dark:bg-primary-600 dark:hover:bg-primary-700 dark:focus:ring-primary-800"
                data-testid="btnSubmit"
                type="submit"
              >
                Sign in
              </button>
              <p
                class="text-sm font-light text-gray-500 dark:text-gray-400"
              >
                Don't have an account yet? 
                <a
                  class="font-medium text-primary-600 hover:underline dark:text-primary-500"
                  href="/register"
                >
                  Sign up
                </a>
              </p>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
